agera/src/main/java/com/google/android/agera/CompiledRepository.java

```diff
@@ -142,20 +142,23 @@ protected void observableDeactivated() {
   @Override
   public void update() {
     maybeCancelFlow(concurrentUpdateConfig, true);
-    maybeStartFlow();
+    maybeStartFlow(false);
   }
 
   /**
    * Called on the worker looper thread. Starts the data processing flow if it's not running. This
    * also cancels the lazily-executed part of the flow if the run state is "paused at lazy".
    */
-  void maybeStartFlow() {
+  void maybeStartFlow(final boolean restartNeeded) {
     synchronized (this) {
       if (runState == IDLE || runState == PAUSED_AT_GO_LAZY) {
         runState = RUNNING;
         lastDirectiveIndex = -1; // this could be pointing at the goLazy directive
-        restartNeeded = false;
+        this.restartNeeded = false;
       } else {
+        if (runState == CANCEL_REQUESTED) {
+          this.restartNeeded |= restartNeeded;
+        }
         return; // flow already running, do not continue.
```

Comment:
Looking at this original comment, I think the race condition is more serious (and frequent) than you think (quoting your keyword "rarely occurring").

The flow that concerns me:

some event -> update() -> maybeStartFlow(false) -> "flow already running, do not continue".

But what if the flow is under a lengthy process (state == RUNNING), but the computation is based on some outdated value retrieved before the event?

I think, with relatively high confidence, that rescheduling restart is unconditional within this else-block. This might cause two or three passes of the flow to stabilize the repository, but we should assume any caching/deduping be handled gracefully by the operators already, anyway, so the cost of this stabilizing isn't that high.

If you agree too, then let's chat to discuss the perfect fix. I'm thinking maybe both `maybeStartFlow()` and `maybeCancelFlow()` can be free of the boolean parameters -- `maybeCancelFlow()` will only set restartNeeded to false, and maybeStartFlow() will only set restartNeeded to true, when appropriate.